Create rule S7121 Calls to c_str() should not implicitly recreate strings or string_views CPP-3435 (#4391)
SonarSource/rspec: rules/S7121/cfamily/rule.adoc

Unnecessary calls to `c_str()` or `data()` reduce readability and performance.

== Why is this an issue?

> In the process of refactoring code from C to C++, it can easily happen that redundant calls to `c_str()` or `data()` on standard strings get introduced when the string could be used directly.

Reviewer: I think we should also cover std::string_view, as this avoid recomputing lenght.

[source,cpp,diff-id=1,diff-type=noncompliant]
----
void takeString(std::string const& dest);

void func(std::string const& src) {
  takeString(src.c_str()); // Noncompliant: redundant copy
}
----

Not only is this unnecessary step a readability issue, but it will force the creation of useless intermediary strings, reducing the performance.

[source,cpp,diff-id=1,diff-type=compliant]
----
void takeString(std::string const& dest);

void func(std::string const& src) {
  takeString(src); // Compliant
}
----

=== Exception

This rule does not raise when explicitly creating a string from `c_str()`.

[source,cpp]
----
void takeString(std::string const& dest);

void func(std::string const& src) {
  takeString(std::string(src.c_str())); // Compliant by exception
}
----

This operation could be done on purpose in the rare case where `src` is a string containing embedded null-terminators, in order to only keep the content up to the first null-terminator in `dest`.

> 

Reviewer: I would add to link to the rule, as for me it is related: 
https://sonarsource.github.io/rspec/#/rspec/S6231

Unnecessary calls to `c_str()` or `data()` reduce readability and performance.

== Why is this an issue?

In the process of refactoring code from C to C++, it can easily happen that redundant calls to `c_str()`

> or `data()` on standard strings get introduced when the original string could be used directly.

Reviewer: ```suggestion
or `data()` on standard strings get introduced when the original object could be used directly.
```

[source,cpp,diff-id=1,diff-type=noncompliant]
----
void takeString(std::string const& dest);

void takeStringView(std::string_view dest);

void func(std::string const& src) {
  takeString(src.c_str()); // Noncompliant: redundant copy
  takeStringView(src.c_str()); // Noncompliant: redundant crossing of the content
  takeStringView(src.data());  // Noncompliant: redundant crossing of the content
}
----

Not only is this unnecessary step a readability issue, but it will force the creation of useless intermediary strings,

> or force to cross the content of a string to search for the null-terminator, reducing the performance.

Reviewer 2: I don't think this is the right term here...
```suggestion
or require searching the content of a string for the null-terminator, reducing the performance.
```

> and force to cross the content of a string to search for the null-terminator, reducing the performance.

Reviewer: ```suggestion
or force to cross the content of a string to search for the null-terminator, reducing the performance.
```
